tests/jerry/unusual-javascript-constructs.js

```diff
@@ -0,0 +1,39 @@
+// Copyright JS Foundation and other contributors, http://js.foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+var x = [0];
+assert (x == x);
+assert (x == !x);
+assert (Array(3) == ",,");
+
+assert (typeof NaN == "number");
+assert (NaN != NaN);
+assert (typeof null == "object");
+assert (!(null instanceof Object));
+assert (!("string" instanceof String));
+
+assert (9999999999999999 == 10000000000000000);
+assert (0.1 + 0.2 != 0.3);
+assert (!(Math.max() > Math.min()));
+assert (Math.max() < Math.min());
+
+assert ([] + [] == "");
+assert ([] + {} == "[object Object]");
+//assert({} + [] == 0); // FIXME: assertion failed, it should be 0, but it is [object Object]
+//assert({} + {} == NaN); // // FIXME: assertion failed, it should be NaN, but it is [object Object][object Object]
```

Comment:
Please create new issues for these two ones! It probably worth to add the link of the issues here as well.

Reply:
There is **no need** to add new issues and also the _assertions are wrong_.

The correct asserts would look like:

```js
assert (eval ("{} + []") == 0)
assert (isNaN (eval ("{} + {}"))); // isNaN should be used instead of == NaN
```
The reason behind this is how `{}` is being parsed.
- As a block (That's why `{} + []` evaluates to 0, since `{}` evaluates to an **empty block**, and `+[]` gives **0** as a result)
- As an object literal (putting this check in an assert the `{}` becomes a part of an **expression** so it's interpreted as an **object**)

Also these test cases should be added with a bit modification:

```js
assert ({} + [] == "[object Object]")
assert ({} + {} == "[object Object][object Object]");
```

Reply:
This is strange. I've just built jerry and the interactive shell gives me:
```
jerry> {} + {}
NaN
```

That's not `[object Object][object Object]`.

Moreover, the assert couldn't be passing anyway as comparing anything to NaN (NaN included) gives false.

Reply:
(To clarify: my above comment "This is strange" is not a response to @rerobika but a reaction to the commit itself. We must have been typing our reviews in parallel.)

Reply:
@akosthekiss That's the expected behavior. Since you are using the CLI, it evaluates the given line, so the output is correct. (same for the other test case)

``` js
jerry> {} + [] // parse {} as an empty block
0
jerry> print ( {} + [] ) // parse {} as an object literal ( now {} is a part of an expression)
[object Object]
undefined
```